pkg/tempo/promsdprocessor/prom_sd_processor.go

```diff
@@ -0,0 +1,216 @@
+package promsdprocessor
+
+import (
+	"context"
+	"net"
+	"strings"
+	"sync"
+
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/discovery"
+	sd_config "github.com/prometheus/prometheus/discovery/config"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/pkg/relabel"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/component/componenterror"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/pdata"
+)
+
+const (
+	addressLabel = "__address__"
+)
+
+type promServiceDiscoProcessor struct {
+	nextConsumer consumer.TraceConsumer
+	discoveryMgr *discovery.Manager
+
+	relabelConfigs map[string][]*relabel.Config
+	hostLabels     map[string]model.LabelSet
+	mtx            sync.Mutex
+
+	logger log.Logger
+}
+
+func newTraceProcessor(nextConsumer consumer.TraceConsumer, scrapeConfigs []*config.ScrapeConfig) (component.TraceProcessor, error) {
+	logger := log.With(util.Logger, "component", "tempo service disco")
+	mgr := discovery.NewManager(context.Background(), logger, discovery.Name("tempo service disco"))
```

Comment:
If Tempo fails this will be a resource leak. Can you make sure the discovery manager gets shutdown when the processor is shutdown?